Added Google Cloud Spanner metrics receiver skeleton.

ydrozhdzhal opened the pull request:
Added Google Cloud Spanner metrics receiver skeleton.

Commit:
Added Google Cloud Spanner receiver skeleton.

bogdandrutu commented:
Hey @ydrozhdzhal please sign the CLA

ydrozhdzhal commented:
> Hey @ydrozhdzhal please sign the CLA
@bogdandrutu , I am in progress with this. I hope to manage this tomorrow.

ydrozhdzhal commented:
> Hey @ydrozhdzhal please sign the CLA

Done

ydrozhdzhal commented:
@bogdandrutu , we signed CLA

anuraaga commented on receiver/googlecloudspannerreceiver/config.go:
```diff
@@ -0,0 +1,114 @@
+type Config struct {
+	scraperhelper.ScraperControllerSettings `mapstructure:",squash"`
+
+	TopMetricsQueryMaxRows int       `mapstructure:"top_metrics_query_max_rows"`
```
How about using a struct `TopMetricsQuery` with a field inside `max_rows` to allow more options to be added later

ydrozhdzhal replied:
Nice idea. For now we do not expect more options.
We can introduce such structure later if required(more options appear).

anuraaga commented on receiver/googlecloudspannerreceiver/config.go:
```diff
@@ -0,0 +1,114 @@
+	scraperhelper.ScraperControllerSettings `mapstructure:",squash"`
+
+	TopMetricsQueryMaxRows int       `mapstructure:"top_metrics_query_max_rows"`
+	Projects               []Project `mapstructure:"projects"`
```
Is it idiomatic for one collector to be scraping multiple projects like this? I figured it's more idiomatic if each project would have a collector, that may export their metrics in OTLP to a collector in a monitoring account if cross-project is needed.

This is fine if it seems conventional for GCP though

ydrozhdzhal replied:
Yes, I agree with you, but:

- current configuration allows to reach the same result
- our use cases require to have one collector for multiple projects with single receiver.

anuraaga commented on receiver/googlecloudspannerreceiver/config.go:
```diff
@@ -0,0 +1,114 @@
+
+type Project struct {
+	ID                string     `mapstructure:"project_id"`
+	ServiceAccountKey string     `mapstructure:"service_account_key"`
```
Not supporting multiple projects would allow using google default credentials I think which could be more user-friendly

ydrozhdzhal replied:
I agree but in our use case it is required to have multiple projects.

punya replied:
For this and https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5240/files#r709960450, could you describe the use cases in some more detail?

ydrozhdzhal replied:
@punya , current configuration options allows you either work with one or multiple projects. Also there was request to us that potential users do not want to have multiple receivers and multiple pipelines configured in scope of one collector instance for Spanner monitoring. They want to have possibility to collect data from all their projects in scope of one receiver instance.

anuraaga commented on receiver/googlecloudspannerreceiver/config.go:
```diff
@@ -0,0 +1,114 @@
+	return nil
+}
+
+func (project Project) Validate() error {
```
Make the sub-validation helpers private instead of public (`validate`)

ydrozhdzhal replied:
Ok, will do

ydrozhdzhal replied:
Done.

Commit:
Made license text consistent with collector repo.

ydrozhdzhal commented:
@anuraaga , I replied to you comments and made changes/updates.

ydrozhdzhal commented:
@balachandr, you asked to add you here also.

ydrozhdzhal commented:
@bogdandrutu what are the next steps? Who can approve workflow for this MR? Do I need to squash(not squash) commits and rebase?

ydrozhdzhal commented:
@bogdandrutu there were failing tests not related to my code updates, so I rebased branch to the latest main and squashed some commits. Can you approve workflow one more time? Will it be merged automatically or more actions are required?

ydrozhdzhal commented:
@bogdandrutu there are failing tests not related to my code updates, so I rebased branch to the latest main and squashed some commits but still no luck.

ydrozhdzhal commented:
@bogdandrutu , @anuraaga looks like workflow needs to be approved again, after re-basing no failures in workflow at last.

ydrozhdzhal commented:
@bogdandrutu Who can press the merge button? Also I am not sure about this one failing test in workflow - looks like it is unrelated to my changes.
